chain/network/src/types.rs

```diff
@@ -199,12 +199,92 @@ pub struct HandshakeV2 {
     pub edge_info: EdgeInfo,
 }
 
-impl From<HandshakeV2> for Handshake {
-    fn from(handshake_old: HandshakeV2) -> Self {
+impl HandshakeV2 {
+    pub fn new(
+        peer_id: PeerId,
+        target_peer_id: PeerId,
+        listen_port: Option<u16>,
+        chain_info: PeerChainInfo,
+        edge_info: EdgeInfo,
+    ) -> Self {
         Self {
-            // In the transition to the new version, we care about the oldest supported version
-            // by the other node.
-            version: handshake_old.oldest_supported_version,
+            version: PROTOCOL_VERSION,
+            oldest_supported_version: OLDEST_BACKWARD_COMPATIBLE_PROTOCOL_VERSION,
+            peer_id,
+            target_peer_id,
+            listen_port,
+            chain_info,
+            edge_info,
+        }
+    }
+}
+
+/// Struct describing the layout for HandshakeV2.
+/// It is used to automatically derive BorshDeserialize.
+#[derive(BorshDeserialize)]
+pub struct HandshakeV2AutoDes {
+    pub version: u32,
+    pub oldest_supported_version: u32,
+    pub peer_id: PeerId,
+    pub target_peer_id: PeerId,
+    pub listen_port: Option<u16>,
+    pub chain_info: PeerChainInfo,
+    pub edge_info: EdgeInfo,
+}
+
+const ERROR_UNEXPECTED_LENGTH_OF_INPUT: &str = "Unexpected length of input";
+
+// Use custom deserializer for HandshakeV2. Try to read version of the other peer from the header.
+// If the version is supported then fallback to standard deserializer,
+// otherwise return HandshakeV2 with current parsed versions and default values in other fields.
+impl BorshDeserialize for HandshakeV2 {
+    fn deserialize(buf: &mut &[u8]) -> std::io::Result<Self> {
+        // Detect the current and oldest supported version from the header
+        if buf.len() < 8 {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                ERROR_UNEXPECTED_LENGTH_OF_INPUT,
+            ));
+        }
+
+        let version = u32::from_le_bytes(buf[..4].try_into().unwrap());
+        let oldest_supported_version = u32::from_le_bytes(buf[4..8].try_into().unwrap());
+
+        if OLDEST_BACKWARD_COMPATIBLE_PROTOCOL_VERSION <= version && version <= PROTOCOL_VERSION {
+            // If we support this version, then try to deserialize with custom deserializer
+            match version {
+                _ => HandshakeV2AutoDes::deserialize(buf).map(Into::into),
+            }
+        } else {
+            // Consume the remainder of the buffer
+            // NOTE: It is important that HandshakeV2 only appears in the end of a message
+            // to be deserialized properly.
+            *buf = &buf[buf.len()..];
+
+            // If we don't support this version, create a mock HandshakeV2 with version values,
+            // and let `PeerActor` send message:
```

Comment:
Here we convert a handshake with not supported version into a HandshakeFailure

https://github.com/nearprotocol/nearcore/blob/ca592796b005702ddb46a3aff051ef52450353ec/chain/network/src/peer.rs#L670

Notice that prior this change, the client will fail to deserialize the message and would not respond to such input. The contract of deserializer, is that if we are able to deserialize the HandshakeV2 we do it as expected, otherwise, if we detect that the version is not supported we create a HandshakeV2 but only populate the fields `version`, `oldest_supported_version` and mock the remaining fields. So basically, fields from HandshakeV2 are only reliable if HandshakeV2.version is supported.

I agree this API feels somehow hacky, but the other alternative to avoid double upgrades and support this feature is relay on custom error `std::io::Result::Error::Custom(VersionNotSupported)`. Probably this is a better (less hacky) approach from a semantic point of view. I'm ok doing it this way.

Reply:
> Probably this is a better (less hacky) approach from a semantic point of view. I'm ok doing it this way.

Yeah I agree

Reply:
done